include/El/blas_like/level1/DiagonalScale.hpp

```diff
@@ -11,67 +11,123 @@
+void DiagonalScale(
+    LeftOrRight side, Orientation orientation,
+    Matrix<T,Device::GPU> const& d, Matrix<T,Device::GPU>& A)
+{
+    EL_DEBUG_CSE;
+    const Int m = A.Height();
+    const Int n = A.Width();
+    const Int lda = A.LDim();
+    const Int incd = 1;
+
+    if (orientation != NORMAL)
+        LogicError("DiagonalScale: Only NORMAL mode supported on GPU");
+
+    cublas::Dgmm(side, m, n,
+                 A.Buffer(), lda,
+                 d.Buffer(), incd,
+                 A.Buffer(), lda);
+}
+
+template <typename T, typename, typename>
+void DiagonalScale(
+    LeftOrRight, Orientation,
+    Matrix<T,Device::GPU> const&, Matrix<T,Device::GPU>&)
+{
+    LogicError("DiagonalScale: Bad device type.");
+}
+
+#endif // HYDROGEN_HAVE_CUDA
+
 template<typename TDiag,typename T>
-void DiagonalScale
-( LeftOrRight side,
-  Orientation orientation,
-  const Matrix<TDiag>& d,
-        Matrix<T>& A )
+void DiagonalScale(
+    LeftOrRight side,
+    Orientation orientation,
+    Matrix<TDiag> const& d,
+    Matrix<T>& A)
 {
-    EL_DEBUG_CSE
+    EL_DEBUG_CSE;
     const Int m = A.Height();
     const Int n = A.Width();
-    const bool conj = ( orientation == ADJOINT );
-    if( side == LEFT )
+    const bool conj = (orientation == ADJOINT);
+    if (side == LEFT)
     {
-        EL_DEBUG_ONLY(
-          if( d.Height() != m )
-              LogicError("Invalid left diagonal scaling dimension");
-        )
-        for( Int i=0; i<m; ++i )
+#ifndef EL_RELEASE
+        if (d.Height() != m)
+            LogicError("Invalid left diagonal scaling dimension");
+#endif
+        for(Int i=0; i<m; ++i)
         {
-            const T delta = ( conj ? Conj(d(i)) : d(i) );
-            for( Int j=0; j<n; ++j )
+            const T delta = (conj ? Conj(d(i)) : d(i));
+            for(Int j=0; j<n; ++j)
                 A(i,j) *= delta;
         }
     }
     else
     {
-        EL_DEBUG_ONLY(
-          if( d.Height() != n )
-              LogicError("Invalid right diagonal scaling dimension");
-        )
-        for( Int j=0; j<n; ++j )
+#ifndef EL_RELEASE
+        if (d.Height() != n)
+            LogicError("Invalid right diagonal scaling dimension");
+#endif
+        for(Int j=0; j<n; ++j)
         {
-            const T delta = ( conj ? Conj(d(j)) : d(j) );
-            for( Int i=0; i<m; ++i )
+            const T delta = (conj ? Conj(d(j)) : d(j));
+            for(Int i=0; i<m; ++i)
                 A(i,j) *= delta;
         }
     }
 }
 
-template<typename TDiag,typename T,Dist U,Dist V,DistWrap wrapType>
-void DiagonalScale
-( LeftOrRight side,
-  Orientation orientation,
-  const AbstractDistMatrix<TDiag>& dPre,
-        DistMatrix<T,U,V,wrapType>& A )
+template <typename T>
+void DiagonalScale(LeftOrRight side,
+                   Orientation orientation,
+                   AbstractMatrix<T> const& d,
+                   AbstractMatrix<T>& A)
 {
-    EL_DEBUG_CSE
-    if( wrapType == ELEMENT )
+    switch (A.GetDevice())
```

Comment:
Do we want to catch the case where the user provides matrices with different devices?

Reply:
Oh, probably.